news.d/cnews.d/rna/man/readnews.1
.TH READNEWS 1
.SH NAME
news, readnews \- read news articles
.SH SYNOPSIS
.B readnews
.RB [ -n
newsgroups]
.RB [ -i ]
.RB [ -clpC ]
.RB [ -s [ -+?
.RI [ group ]]]
.RB [ -u
messageid]
.SH DESCRIPTION
.I Readnews
without arguments enters command mode,
which allows printing of unread articles.
This is the normal way of using
.IR readnews .
.P
.I Readnews
maintains a
.I .newsrc
file in the user's home directory that specifies
all news articles already read.
It is updated at the end of each reading session.
.P
Some useful functions are available which don't use command mode.
The flags for these are:
.TP
.B -c
Check if there is news, and if so print `You have news.'.
A line `readnews -c' is usually placed in the system
.I .profile
.RB (  /etc/profile ).
.TP
.B -C
Check if there is news, and print the groups and number of
articles in each group to be read.
.TP
.B -l
List the titles of available news articles.
.TP
.B -p
Print all articles on standard output,
and update
.IR newsrc .
.TP
.B -s
Print the newsgroup subscription list.
.TP
.BI -s+ " group"
Add
.I group
to the subscription list.
.TP
.BI -s- " group"
Subtract
.I group
from the subscription list.
.TP
.B -s?
List currently active newsgroups.
.P
The remaining flags determine article selection,
and may also appear in the
.I .newsrc
file.
Options may be specified in the
.I .newsrc
file by entering lines prefixed with the word `options',
followed by the options arguments.
This is most useful with the
.B -n
flag, specifying the usual groups one wishes to subscribe to.
.TP
\fB-n \fInewsgroups\fR
Select all articles belonging to
.IR newsgroups .
.I newsgroups
is a comma separated list of newsgroup names.
The character `!' may be used to exclude certain groups,
and the word `all' can be used to match any group.
e.g. `-n all,!net.jokes'
.TP
.B -i
Ignore
.I .newsrc
file. It is not read or updated.
This allows selection of articles that have already been read.
.TP
\fB-u \fImessageid\fR
Unsubscribe to followup articles referring to
.IR messageid .
(This flag is usually only placed in the
.I .newsrc
file as a result of the `u' command.)
.SH COMMANDS
This section details the commands available when
.I readnews
is in command mode (no
.B -clpsC
arguments).
The simplest way of using this mode, is to enter RETURN after every
prompt.
This will present to the user, a short heading for an article, then a prompt.
Typing RETURN again will print the article body.
Typing RETURN yet again will print the next heading, and so on.
If having read the heading, you don't wish to read the article, you may
type `n' (or `+' or ';') which will take you directly to the next heading.
.P
.P
An article is treated as having been read, if either you have seen
the article body, or typed `n' to skip over it.
.P
A number of commands operate on the `current' article.
This is defined as the article whose header you have most recently seen.
.P
The commands to read news are:
.TP
.B RETURN
Either print the current article,
or go to the next article and print its header.
.TP
\fBn\fR or \fB+\fR or \fB;\fR
Go to the next article and print its header.
.TP
.B .
Print the current article.
.TP
.B -
Go back to the previous article. This is a toggle, typing it
twice returns you to the original article.
.TP
.I number
Go to the article
.I number
in the current newsgroup.
Like the `-' command,
you always return to the original article
after reading the selected article.
.TP
\fBs \fR[\fIfile\fR]
Save the current article, either in the specified file, or
in
.BR $HOME/articles .
.TP
.B h
Print the current header (slightly more verbose than normal header).
.TP
.B H
Print the current header in full (very verbose).
.TP
\fBN \fR[\fInewsgroup\fR]
Go to the next newsgroup, or to the specified newsgroup.
.TP
.B u
Unsubscribe from all further followup articles on this topic.
.TP
.B U
Unsubscribe from this newsgroup, and go to the next newsgroup.
.TP
\fB!\fIcommand\fB
Shell escape.
.I Command
is executed.
If 
.I command
is `!'
the last escape command is executed.
.TP
\fBq\fR or \fBEOT\fR
Quit.
The
.I .newsrc
file will be updated provided the flag
.B -i
was not specified.
.TP
.B x
Exit.
.I .newsrc
is left unchanged (as if no articles had been read).
.TP
.B DEL
An interrupt will cause
.I readnews
to terminate its current activity and return to command mode.
An interrupt in command mode will cause `Interrupt' to be printed,
and a subsequent interrupt will cause immediate exit (as in the `x' command).
.P
Some commands are available to send/reply or cancel news articles:
.TP
.B c
Cancel article. Only the author, or news administrator can do this.
.IR postnews (1)
is called to do the actual cancelling.
.TP
.B r
Reply to sender of the current article by mail.
.I Readnews
sets up the appropriate headers, and then calls
.IR mail (1)
to send a reply to the sender.
.TP
.B f
Post a followup to the current article.
.I Readnews
sets up the appropriate headers, and then calls
.IR postnews (1)
to post the followup article.
.TP
.B p
Post an article on a new topic.
.IR postnews (1)
is called to post the new article.
.TP
\fBm \fIperson\fB
Mail to
.IR person .
.P
When replying by mail, or posting an article, the user
is prompted for certain headers, and then the text of the article or mail
item is entered until a `.' or EOT is entered alone on a line.
Then the article/mail is posted/mailed.
Other commands are available:
.TP
\&\fB.e\fR
Edit the message/article collected so far (see
.IR ed (1)).
The 'To:' or 'cc:' fields may be changed if mailing.
After editing further lines may be appended to the message.
.TP
\&\fB.i\fR
Interpolate
the current news article onto the end of the message.
The interpolated item
is indented by four spaces.
.TP
\&\fB.!\fIcmd\fR or \fB!\fIcmd\fR
Shell escape.
.IR Cmd
is executed.
.TP
.B DEL
Causes posting/mailing to be aborted, and the article entered so far
is saved in
.B $HOME/dead.article
or
.BR $HOME/dead.letter .
.P
In order to permanently resubscribe to a newsgroup denied by `U',
or a series of followups denied by `u' it is necessary to understand
the format of the
.I .newsrc
file.
The
.I .newsrc
file consists of two types of lines:
.TP
option lines
These start with the word `option' and contain the same arguments
as the
.I readnews
command on the command line.
Followups are denied with `option -u <messageid>'.
To resubscribe to further followups, the correct options line must be deleted.
.TP
read newsgroup lines
These have the format <newsgroup>`:' <number>, where
<number> represents the last item number seen in that particular newsgroup.
If the newsgroup has been unsubscribed, the `:' is replaced by a `!'.
To resubscribe the `!' must be changed back to a `:'.
.SH FILES
.ta 24
.nf
$HOME/.newsrc	options and list of previously read articles
%news	where the articles are kept
/usr/lib/news/active	current newsgroups
/usr/lib/news/help	help file
.fi
.SH SEE ALSO
postnews(1), mail(1), ed(1), uusend(8), uurec(8).
.SH BUGS
.I Readnews
with the
.B -c
flag may say "You have news.", when the available article is a unsubscribed
followup article.
.P
You may see followups, even if you have used the `u' command.
This is because many sites have faulty news programs, which do
not follow the correct protocol, or the sender did not use the `r' command.
.SH AUTHOR
Michael Rourke, University of N.S.W (decvax!mulga!michaelr:elecvax)


local.d/news.d/src/readnews.c
/*
 * readnews - read news articles.
 */

#ifndef lint
static char	*SccsId = "@(#)readnews.c	2.19	8/28/84";
#endif !lint

#include "rparams.h"

/*
 * readnews - article reading program
 */

#ifndef SYSBUF
char	SYSBUF[BUFSIZ];	/* to buffer std out */
#endif

#define OPTION	0	/* pick up an option string */
#define STRING	1	/* pick up a string of arguments */

int onsig();

/*
 *	Authors:
 *		Matt Glickman	ucbvax!glickman
 *		Mark Horton	cbosg!mark
 *		Stephen Daniels	duke!swd
 *		Tom Truscott	duke!trt
 */

main(argc, argv)
int	argc;
register char	**argv;
{
	register char	*ptr;	/* pointer to rest of buffer		*/
	char	*user, *home;
	int	optflag = FALSE, space = FALSE;
	struct utsname ubuf;
	char	*myrc;

	/* set up defaults and initialize. */
	pathinit();
	mode = UNKNOWN;
	header.title[0] = header.nbuf[0] = '\0';
	titlebuf[0] = coptbuf[0] = datebuf[0] = '\0';
	uname(&ubuf);
	strcpy(FULLSYSNAME, ubuf.nodename);

	savmask = umask(N_UMASK);	/* set up mask */
	uid = getuid();
	gid = getgid();
	duid = 0;
	dgid = 0;

#ifndef V6
#ifndef SHELL
	if ((SHELL = getenv("SHELL")) == NULL)
		SHELL = "/bin/sh";
#endif
#ifndef IHCC
	/*
	 * IHCC does not allow use of $LOGNAME to prevent forgery.
	 * Note that this shouldn't matter in readnews, since inews
	 * does all the actual posting of news.
	 */
	if ((user = getenv("USER")) == NULL)
		user = getenv("LOGNAME");
	if ((home = getenv("HOME")) == NULL)
		home = getenv("LOGDIR");
#endif
	if (user == NULL || home == NULL)
		getuser();
	else {
		username = AllocCpy(user);
		strcpy(header.path, username);
		userhome = AllocCpy(home);
	}

	getuser();
	if (!(MAILER = getenv("MAILER")))
		MAILER = "mail";	/* was /bin/mail */

#ifdef PAGE
	if (myrc = getenv("PAGER"))
		strcpy(PAGER, myrc);
	else
# ifdef IHCC
		sprintf(PAGER,"%s/bin/%s",logdir(HOME),PAGE);
# else
		strcpy(PAGER, PAGE);
# endif
#else
	strcpy(PAGER, "");
#endif

	if (ptr = getenv("NEWSOPTS"))
		strcpy(rcbuf, ptr);
	else
		*rcbuf = '\0';
	if (*rcbuf) {
		strcat(rcbuf, " \1");
		ptr = rcbuf;
		while (*++ptr)
			if (isspace(*ptr))
				*ptr = '\0';
		for (ptr = rcbuf; ; ptr++) {
			if (!*ptr)
				continue;
			if (*ptr == '\1')
				break;
			if (++line > LINES)
				xerror("Too many options.");
			if ((rcline[line] = malloc((unsigned)(strlen(ptr) + 1))) == NULL)
				xerror("Not enough memory.");
			argvrc[line] = rcline[line];
			strcpy(rcline[line], ptr);
			while (*ptr)
				ptr++;
		}
	}
#else
	getuser();
#endif
	myrc = getenv("NEWSRC");
	if (myrc == NULL) {
		myrc = NEWSRC;
		sprintf(newsrc, "%s/%s", userhome, myrc);
	} else {
		strcpy(newsrc, myrc);
	}
	if (access(newsrc, 0))
		newrc(newsrc);
	if ((rcfp = fopen(newsrc, "r")) != NULL) {
		rcreadok = FALSE;
		while (fgets(rcbuf, LBUFLEN, rcfp) != NULL) {
			if (!(space = isspace(*rcbuf)))
				optflag = FALSE;
			if (!strncmp(rcbuf, "options ", 8))
				optflag = TRUE;
			if (optflag) {
				strcat(rcbuf, "\1");
				if (space)
					ptr = rcbuf - 1;
				else
					ptr = &rcbuf[7];
				while (*++ptr)
					if (isspace(*ptr))
						*ptr = '\0';
				if (space)
					ptr = rcbuf;
				else
					ptr = &rcbuf[8];
				for (; ; ptr++) {
					if (!*ptr)
						continue;
					if (*ptr == '\1')
						break;
					if (++line > LINES)
						xerror("Too many options.");
					if ((rcline[line] = malloc((unsigned)(strlen(ptr) + 1))) == NULL)
						xerror("Not enough memory.");
					argvrc[line] = rcline[line];
					strcpy(rcline[line], ptr);
					while (*ptr)
						ptr++;
				}
			}
		}
		fclose(rcfp);
		rcreadok = TRUE;
	}
	if (line != -1) {
#ifdef DEBUG
		for (i = 0; i <= line; i++)
			fprintf(stderr, "options:  %s\n", rcline[i]);
#endif
		process(line + 2, argvrc);
		do {
#ifdef DEBUG
			fprintf(stderr, "Freeing %d\n", line);
#endif
			free(rcline[line]);
		} while (line--);
	}

	argv++;
	strcat(header.nbuf, ADMSUB);
	ngcat(header.nbuf);
	process(argc, argv);
	if (!nflag) {
		strcpy(header.nbuf, DFLTSUB);
		ngcat(header.nbuf);
		strcat(header.nbuf, ADMSUB);
		ngcat(header.nbuf);
	}
	if (*header.nbuf)
		lcase(header.nbuf);
	makehimask(header.nbuf, "junk");
	makehimask(header.nbuf, "control");
	makehimask(header.nbuf, "test");

	setbuf(stdout, SYSBUF);
	sigtrap = FALSE;	/* true if a signal has been caught */
	if (!pflag && !lflag && !eflag) {
		signal(SIGQUIT, SIG_IGN);
		signal(SIGHUP, onsig);
		signal(SIGINT, onsig);
		signal(SIGPIPE, onsig);
	}

	/*
	 * ALL of the command line has now been processed. (!)
	 */

	if (!*header.nbuf)
		ngcat(strcpy(header.nbuf, DFLTSUB));
	if (sflag) {
		ngdel(header.nbuf);
		printf("Subscription list:  %s\n", header.nbuf);
		xxit(0);
	}
	if (xflag)
		line = -1;
	rcfp = xfopen(newsrc, "r");
	while (fgets(rcbuf, LBUFLEN, rcfp) != NULL) {
		if (!nstrip(rcbuf))
			xerror(".newsrc line too long");
		if (++line >= LINES)
			xerror("Too many .newsrc lines");
		if ((rcline[line] = malloc((unsigned)(strlen(rcbuf) + 1))) == NULL)
			xerror("Not enough memory");
		strcpy(rcline[line], rcbuf);
	}
	fclose(rcfp);

	if (sigtrap) {
		if (sigtrap == SIGHUP || !rcreadok)
			xxit(0);
		fprintf(stdout, "Abort (n)?  ");
		fflush(stdout);
		if (gets(bfr) == NULL || *bfr == 'y' || *bfr == 'Y')
			xxit(0);
		sigtrap = FALSE;
	}
#ifdef SORTACTIVE
	sortactive();
#endif
	actfp = xfopen(ACTIVE, "r");

#ifdef DEBUG
	fprintf(stderr, "header.nbuf = %s\n", header.nbuf);
#endif
	switch (mode) {
	case UNKNOWN:
		readr();
		break;
#ifdef TMAIL
	case MAIL:
		Mail();
		break;
#endif
	}
	fflush(stdout);
	if (xflag || lflag || tflag)
		xxit(0);
	if (*groupdir && mode != MAIL)
		updaterc();
	writeoutrc();
	xxit(0);

	/* Camel, R.O.H. */
}

/*
 * Write out the .newsrc file.  We sort it into "active" file order,
 * for speed in future invocations, and to get rid of junk.
 */
writeoutrc()
{
	FILE *wrcfp, *afp;
	char aline[BUFLEN];
	register int i, c;
	register char *p;

	if (!rcreadok)
		return;
#ifdef VMS
	unlink(newsrc);
#endif

	wrcfp = xfopen(newsrc, "w");
	afp = xfopen(ACTIVE, "r");

	/* Write out options line, continuations, and comments. */
	for (i=0;rcline[i];i++) {
		c = rcline[i][0];
		if (c != 'o' && c != '#' && c != ' ' && c != '\t')
			break;
		if (c == 'o' && strncmp(rcline[i], "options", 7) != 0)
			break;
		fprintf(wrcfp, "%s\n", rcline[i]);
	}

	/* For each newsgroup in active, find that newsrc line and write it out. */
	while (fgets(aline, sizeof aline, afp)) {
		p = index(aline, ' ');
		if (p)
			*p = 0;
		i = findrcline(aline);
		if (i >= 0)
			fprintf(wrcfp, "%s\n", rcline[i]);
	}
	fclose(wrcfp);
	fclose(afp);
}

/*
 * Forbid newsgroup ng, unless he asked for it in nbuf.
 */
makehimask(nbuf, ng)
char *nbuf, *ng;
{
	if (!findex(nbuf, ng)) {
		ngcat(nbuf);
		strcat(nbuf, "!");
		strcat(nbuf, ng);
		ngcat(nbuf);
	}
}

/*
 * Return true if the string searchfor is in string, but not if preceeded by !.
 */
findex(string, searchfor)
char *string, *searchfor;
{
	register char first;
	register char *p;

	first = *searchfor;
	for (p=index(string, first); p; p = index(p+1, first)) {
		if (((p==string) || (p[-1]!='!')) && strncmp(p, searchfor, strlen(searchfor)) == 0)
			return TRUE;
	}
	return FALSE;
}
